### ISSUE TITLE
Adding ASVS requirement for CWE-470: Unsafe Reflection

### ISSUE THREAD

**Issue by streichsbaer** [OWASP/ASVS#1368]
Hello ASVS team!

I've been reviewing ASVS categories against a list of specific SAST rules to understand what coverage exists and how low-level rules map against ASVS requirements. So far the correlation is quite good and I've only come across one requirement that I think is missing in the current bleeding edge (5.0), and that is [CWE-470: Unsafe Reflection](https://cwe.mitre.org/data/definitions/470.html).

I propose adding it as a new requirement to [V5.1 Input Validation](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x13-V5-Validation-Sanitization-Encoding.md#v51-input-validation), with the following details:

- Requirement Id: 5.1.7
- Requirement description: Verify that allow lists or indirect selection are applied when using reflection to load classes or code based on user input.
- Requirement Verification Level: 1
- Requirement CWE Mapping: 470

Looking forward to your thoughts on this and whether you agree, or think it's already sufficiently covered in another requirement.

Thanks,
Stefan

**Comment by elarlang** [comment]
Hi,

for me it seems to be combination of 2 problems which I would like to watch separately:
* input validation - against allowlist
* loading external resources - from allow-listed sources

From input validation perspective - current V5.1.4 is covering allow-list part
* _V5.1.4 Verify that structured data is strongly typed and validated against a defined schema including allowed characters, length and pattern (e.g. credit card numbers, e-mail addresses, telephone numbers, or validating that two related fields are reasonable, such as checking that suburb and zip/postcode match)._

Additionally you can apply "file execution" requirements like:
* _V12.3.1 Verify that user-submitted filename metadata is not used directly by system or framework filesystems and that a URL API is used to protect against path traversal._
* _V12.3.2 Verify that user-submitted filename metadata is validated or ignored to prevent the disclosure, creation, updating or removal of local files (LFI)._
* _V12.3.3 Verify that user-submitted filename metadata is validated or ignored to prevent the disclosure or execution of remote files via Remote File Inclusion (RFI) or Server-side Request Forgery (SSRF) attacks._
* _V12.3.5 Verify that untrusted file metadata is not used directly with system API or libraries, to protect against OS command injection._
* _V12.3.6 Verify that the application does not include and execute functionality from untrusted sources, such as unverified content distribution networks, JavaScript libraries, node npm libraries, or server-side DLLs._

Related opened issue on loading external resources (requirements 10.3.2, 14.2.3):
* https://github.com/OWASP/ASVS/issues/1293

Also a bit related opened issue for documenting all allow-listed external source:
* https://github.com/OWASP/ASVS/issues/993

**Comment by streichsbaer** [comment]
Thanks for the quick response, @elarlang!

I was looking at [V5.1.4](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x13-V5-Validation-Sanitization-Encoding.md#v51-input-validation), however, it doesn't work for unsafe reflection as you may have access to other functions that would still pass the validation. 

Take for example the requirement _[V5.1.5](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x13-V5-Validation-Sanitization-Encoding.md#v51-input-validation) Verify that URL redirects and forwards only allow destinations which appear on an allow list, or show a warning when redirecting to potentially untrusted content._
It is specifically asking for a "domain" specific allowlist because the requirement 5.1.4 would not sufficiently cover this attack, e.g. applying a regex (for illustration purposes) such as `/^https:\/\/[a-zA-Z0-9.-]*/`. 

The same logic applies to the unsafe reflection, where you may have a bunch of functions but only want users to access the `ReadCommand` and `ListCommand` functions, but not the `DeleteCommand` and `CreateCommand` functions.

Regarding the file execution/loading external resource requirements, I don't think they directly apply to the unsafe reflection vulnerability class.

Best regards,
Stefan

**Comment by elarlang** [comment]
I guess I don't get the actual problem here or I'm just thinking quite a different way on the topic - based on your example, isn't more access control issue then? Something we should not solve with input validation?

**Comment by streichsbaer** [comment]
According to the potential mitigations in the [CWE reference](https://cwe.mitre.org/data/definitions/470.html), the following solutions are recommended.
![image](https://user-images.githubusercontent.com/5006784/191004145-cbadefdd-e1ae-4564-95f9-1fee8e4893f8.png)

Happy to hear other people's thoughts on where this would better fit as a requirement.

I do agree that chapter 5 can do with some improvements, for example:

1. _5.5.2 Verify that the application correctly restricts XML parsers to only use the most restrictive configuration possible and to ensure that unsafe features such as resolving external entities are disabled to prevent XML eXternal Entity (XXE) attacks._: This is part of the _V5.5 Deserialization Prevention_ category, but XXE is not related to deserialization.

**Comment by elarlang** [comment]
Just the fact that it has mapping in the CWE database, does not give guarantee that it is not deprecated.

For me, the problem description here and also in CWE seems deprecated as well - from the "hyper-dynamic-php" era problems, where entire application functionality and user-data (files) maybe kept together in the same directory and/or application included dynamically files and functionality without properly checking permissions.

This assumption is matching with "Potential Mitigations" section - it indicates, that architecture for this kind of application is so wrong, that you just need to rewrite it.

In general I think the issue is covered with access control requirement 4.1.3:
_V4.1.3 Verify that the principle of least privilege exists - users should only be able to access functions, data files, URLs, controllers, services, and other resources, for which they possess specific authorization. This implies protection against spoofing and elevation of privilege._

If you still think it is up-to-date problem and needs separate requirement in ASVS, please provide some clear attack and mitigation example.

For other not-related problems please open separate issue(s).

**Comment by streichsbaer** [comment]
I agree with you that _V4.1.3_ indeed fits this CWE much better than the _V5.1_ category.

Interestingly, in the OWASP Top 10 2021, the CWE-470 is parked under A03-Injection, which could correlate with _V5.3 Output Encoding and Injection Prevention_.

Has there been any discussion about referencing multiple CWEs to ASVS requirements? In my opinion, this would make sense as the mapping of one CWE for a given ASVS requirement is not always sufficient. Even more so when considering that there are often multiple ASVS requirements mapping to the same CWE.

I guess there are 3 different questions now that need to be answered.
1. Can we cover "unsafe reflection" under _V4.1.3_?
2. Does it make sense to add multiple CWEs to a given ASVS requirement?
3. Should _V5.5.2_ be moved out of Deserialization Prevention and into either _V14.1.2_ or _14.1.6_?

Based on your answers, I can also create separate issues for points 2 and 3.
Not sure what the process is regarding point 1.

**Comment by jmanico** [comment]
This is also very similar to the deserialization issues we saw in 2015+

Untrusted data should not drive object creation, in general.

**Comment by elarlang** [comment]
1. based on this issue "unsafe reflection" is "anti-pattern" from ancient times and I don't see any need to cover that with this wording. The principle is covered with 4.1.3 wording.
2. I see that there is need for many-to-many relation with CWE mapping, but realistically I can not see from where to take resources for creating and later maintaining this kind of mapping inside ASVS. There is separate mapping project for that in OWASP, so at the moment the answer is: we don't do it.
3. Like Jim said, there is deserialization component involved when you "translate" XML to objects. So, it's not wrong at the moment. On the other hand, additionally to current 5.5.2 there are some other issues with processing with XML, which is not covered at the moment - like XML bomb a'ka "billion laughs attack", allowing XIncludes etc.
